### PR TITLE
Add default task repeater

### LINKED ISSUE #187: Client should retry on network issues or API errors by simple configuration
### Expected behavior
On network issues or 500 errors from API it should retry request
### Actual behavior
It rejects with error and caller has to implement retry logic

Possible implementation for default retry: https://gist.github.com/butenkor/93098cdd1c2185ad7102b10d9b30cd7e

### PR DESCRIPTION
**Pull request by LEQADA**
Fixes #187

- [x] commit messages are commitizen-friendly
- [x] code is unit tested
- [x] code is integration tested
- [x] public code is documented
- [x] code is reviewed by at least one person
- [x] code satisfies linter rules

### CODE REVIEW

**Comment by LEQADA** [comment]
@Siilwyn @hisabimbola please review

File: src/coffee/client.coffee [hunk 1]
```diff
@@ -220,9 +220,8 @@ class SphereClient
   #
   # options - An {Object} to configure the client
   constructor: (options = {}) ->
-
     # Private: instance of a {TaskQueue}
-    @_task = options.task or new TaskQueue
+    @_task = options.task or new RepeaterTaskQueue {}, { attempts: 50, timeout: 200, timeoutType: 'v' }
```
Comment (LEQADA) [inline src/coffee/client.coffee]: Actually 50 attempts is not so bad because we use [Exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) approach for timeout.
A would create disableRepeater flag to turn repeater off. Is it ok?

File: src/spec/repeater-task-queue.spec.coffee [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+  it 'should finally resolve after three tries', (done) ->
+
+    callsMap = {
+      0: { statusCode: 500, message: 'ETIMEDOUT' }
+      1: { statusCode: 500, message: 'ETIMEDOUT' }
+      2: { statusCode: 200, message: 'success' }
+    }
+    callCount = 0
+    spyOn(@client._rest, 'GET').andCallFake (resource, callback) ->
```
Comment (LEQADA) [inline src/spec/repeater-task-queue.spec.coffee]: Actually there was sinon stubs before but I didn't want to introduce new dependency only for 2-3 tests.

File: src/coffee/client.coffee (hunk 1 above)
Comment (Siilwyn) [inline src/coffee/client.coffee]: Prefer letting them customise the number so there will be one option/flag: to modify the amount of retries and to disable it completely. So setting that option to 0 would disable it.

File: src/spec/repeater-task-queue.spec.coffee (hunk 2 above)
Comment (Siilwyn) [inline src/spec/repeater-task-queue.spec.coffee]: I would totally agree if we're talking about a non-dev dependency. I'm okay with the current code on the other hand adding one development dependency doesn't have many cons...

File: src/coffee/client.coffee (hunk 1 above)
Comment (LEQADA) [inline src/coffee/client.coffee]: You can pass to client constructor your own Repeater to customise `attempts`, `timeout` and `timeoutType`.

Comment (hisabimbola) [inline src/coffee/client.coffee]: I do not get what you mean by _Exponentials backoff_ and how it explains the fact that we want to do 50 retries on a request in 200 ms.

Comment (hisabimbola) [inline src/coffee/client.coffee]: > You can pass to client constructor your own Repeater to customise attempts, timeout and timeoutType.

What @Siilwyn and I are trying to say is that should be possible to customise this values.

Value should not be hard-coded.

File: src/spec/repeater-task-queue.spec.coffee (hunk 2 above)
Comment (hisabimbola) [inline src/spec/repeater-task-queue.spec.coffee]: It's a devDependency and jasmine is not know for stubbing. The cons are not much but I have no strong opinion on this, so fine with me this way.

File: src/coffee/client.coffee (hunk 1 above)
Comment (LEQADA) [inline src/coffee/client.coffee]: @hisabimbola I think that there should be some default values because a lot of projects already uses this SDK and we cannot force them to change Client creation code. 
But as I said you can pass there your own repeater.

Comment (hisabimbola) [inline src/coffee/client.coffee]: I am not saying we should not have default values at all. I am saying the user should be able to customize this value.

Similar logic done [here](https://github.com/sphereio/sphere-node-sdk/blob/master/src/coffee/client.coffee#L228)

 and if not passed in by the user when instantiating the class, we use the defaults to instantiate the class.

Comment (LEQADA) [inline src/coffee/client.coffee]: @hisabimbola please have a look to the last commit

File: src/coffee/repeater-task-queue.coffee [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  'getaddrinfo ENOTFOUND'
+  'Cannot commit on stream id'
+]
+
+# Public: A `RepeaterTaskQueue` adds request repeater on particular response errors
+#
+# `RepeaterTaskQueue` receives two objects as parameter.
+# First object overriding `maxParallel` value of `TaskQueue`
+# Second object contains information about the count of attempts, timeout and timeout type.
+# There are 2 types of `timeoutType`:
```
Comment (PhilippSpo) [inline src/coffee/repeater-task-queue.coffee]: If there are only 2 types, could we also make it a boolean to reduce cognitive load on the developers using this?

Comment (butenkor) [inline src/coffee/repeater-task-queue.coffee]: @PhilippSpo It would require breaking change [here](https://github.com/sphereio/sphere-node-utils/blob/master/src/coffee/helpers/repeater.coffee#L28)

Comment (PhilippSpo) [inline src/coffee/repeater-task-queue.coffee]: Sorry I thought this was new code. Nevermind then. 👍

File: src/coffee/client.coffee [hunk 4]
```diff
@@ -220,9 +226,11 @@ class SphereClient
   #
   # options - An {Object} to configure the client
   constructor: (options = {}) ->
-
     # Private: instance of a {TaskQueue}
-    @_task = options.task or new TaskQueue
+    if options.enableRepeater? and !options.enableRepeater
+      @_task = new TaskQueue
```
Comment (emmenko) [inline src/coffee/client.coffee]: Question: do we really need to keep 2 task queues? Can't we just have e.g: one which can optionally work as repeater but by default is a "normal" task queue?

File: src/coffee/repeater-task-queue.coffee [hunk 5]
```diff
@@ -0,0 +1,73 @@
+{ Repeater } = require 'sphere-node-utils'
+Promise = require 'bluebird'
+util = require 'util'
+TaskQueue = require './task-queue'
+
+# Response messages which will be handled by `RepeaterTaskQueue`
+retryKeywords = [
```
Comment (emmenko) [inline src/coffee/repeater-task-queue.coffee]: Hmm, this looks very fragile. Usually you retry based on the status code. Having hard coded pieces of strings is really not flexible if e.g.: the message changes.
Can we get rid of this list?

File: src/coffee/repeater-task-queue.coffee [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      @_maybeExecute()
+    .done()
+
+
+  _shouldRetry: (error) ->
+    if error and (error.code and util.inspect(error.code).startsWith('5') or error.statusCode and util.inspect(error.statusCode).startsWith('5'))
+      return true
+    retryKeywords.find (str) ->
+      util.inspect(error, depth: null).toUpperCase().indexOf(str.toUpperCase()) > -1
+      return
```
Comment (emmenko) [inline src/coffee/repeater-task-queue.coffee]: I don't like this part (see comment above)

File: src/spec/client/client.spec.coffee [hunk 7]
```diff
@@ -77,6 +77,26 @@ describe 'SphereClient', ->
     expect(=> @client.setMaxParallel(0)).toThrow new Error 'MaxParallel must be a number between 1 and 100'
     expect(=> @client.setMaxParallel(101)).toThrow new Error 'MaxParallel must be a number between 1 and 100'
 
+  it 'should not repeat request on error',(done) ->
```
Comment (emmenko) [inline src/spec/client/client.spec.coffee]: "... if repeater is disabled"

File: src/spec/connect/oauth2.spec.coffee [hunk 8]
```diff
@@ -10,7 +10,7 @@ describe 'OAuth2', ->
     expect(oa._options.host).toBe 'auth.sphere.io'
     expect(oa._options.protocol).toBe 'https'
     expect(oa._options.accessTokenUrl).toBe '/oauth/token'
-    expect(oa._options.timeout).toBe 20000
+    expect(oa._options.timeout).toBe 60000
```
Comment (emmenko) [inline src/spec/connect/oauth2.spec.coffee]: Why??

File: src/spec/integration/products-sync.spec.coffee [hunk 9]
```diff
@@ -248,7 +248,7 @@ describe 'Integration Products Sync', ->
       expect(updated.variants[1].prices[0].country).toBe 'FR'
       done()
     .catch (error) -> done(_.prettify(error))
-  , 10000 # 10sec
+  , 30000 # 30sec
```
Comment (emmenko) [inline src/spec/integration/products-sync.spec.coffee]: Why did you change all of those timeouts?

File: src/coffee/repeater-task-queue.coffee (hunk 6 above)
Comment (emmenko) [inline src/coffee/repeater-task-queue.coffee]: Also, we should make it configurable in which cases the request should be retried.
For example by having a "matching" function that needs to implemented from the user. That way we don't make any assumptions which requests should be retried.

File: src/spec/connect/oauth2.spec.coffee (hunk 8 above)
Comment (LEQADA) [inline src/spec/connect/oauth2.spec.coffee]: Because sometimes it fails on my machine with old timeout

File: src/spec/integration/products-sync.spec.coffee (hunk 9 above)
Comment (LEQADA) [inline src/spec/integration/products-sync.spec.coffee]: Because sometimes it fails on my machine with old timeout

File: src/coffee/repeater-task-queue.coffee (hunk 6 above)
Comment (LEQADA) [inline src/coffee/repeater-task-queue.coffee]: For now it's the best way to do that. Because sometimes there are no error codes at all. I think we can change it in future when we will have more information about this errors.

File: src/coffee/client.coffee (hunk 4 above)
Comment (LEQADA) [inline src/coffee/client.coffee]: It's possible I think. But now it's easier to read and understand.

File: src/coffee/repeater-task-queue.coffee (hunk 5 above)
Comment (dhettler) [inline src/coffee/repeater-task-queue.coffee]: @emmenko Not all of the errors in the list return a proper error code. E.g. in some cases we got
```
"code": "EAI_AGAIN",
"errno": "EAI_AGAIN",
}
```
and ENOTFOUND doesn't even contain these properties.

File: src/spec/client/client.spec.coffee (hunk 7 above)
Comment (LEQADA) [inline src/spec/client/client.spec.coffee]: Done

File: src/coffee/repeater-task-queue.coffee (hunk 5 above)
Comment (butenkor) [inline src/coffee/repeater-task-queue.coffee]: @emmenko Some examples where status code is not defined:

```
{
    "error": "{ [HttpError: getaddrinfo EAI_AGAIN api-at-rackspace.sphere.io:443]\n  message: 'getaddrinfo EAI_AGAIN api-at-rackspace.sphere.io:443',\n  body: \n   { statusCode: undefined,\n     message: 'getaddrinfo EAI_AGAIN api-at-rackspace.sphere.io:443',\n     originalRequest: \n      { endpoint: '/channels?where=key%3D%22orderexport%22',\n        options: \n         { config: { project_key: 'xxx' },\n           host: 'api-at-rackspace.sphere.io',\n           protocol: 'https',\n           timeout: 360000,\n           rejectUnauthorized: true,\n           headers: \n            { 'User-Agent': 'xxx-cli-0.13.4',\n              Authorization: 'Bearer **********' },\n           uri: 'https://api-at-rackspace.sphere.io/xxx' } } },\n  name: 'HttpError',\n  code: undefined }",
    "hostname": "xxx",
    "level": 50,
    "msg": "'order-export' - Unexpected error after '0d 0h 0m 5.297s'",
    "name": "xxx-cli-0.13.4",
    "pid": 1,
    "time": "2016-09-13T18:25:09.745Z",
    "v": 0
}
```

```
{
    "error": "Error: read ECONNRESET\n    at exports._errnoException (util.js:907:11)\n    at TCP.onread (net.js:557:26)",
    "hostname": "xxx",
    "level": 50,
    "msg": "Error details: ",
    "name": "xxx-cli-0.12.9",
    "pid": 1,
    "time": "2016-09-06T00:46:01.203Z",
    "v": 0
}
```

We could make this list as part of constructor which might be overridden if defined in options?

File: src/coffee/repeater-task-queue.coffee (hunk 6 above)
Comment (butenkor) [inline src/coffee/repeater-task-queue.coffee]: @emmenko See comment above (user could optionally define a list of messages to retry). You can always extend ```RepeaterTaskQueue```, override ```_shouldRetry``` and pass it to client instead.

File: src/coffee/client.coffee [hunk 10]
```diff
@@ -220,9 +226,11 @@ class SphereClient
   #
   # options - An {Object} to configure the client
   constructor: (options = {}) ->
-
     # Private: instance of a {TaskQueue}
-    @_task = options.task or new TaskQueue
+    if options.enableRepeater? and !options.enableRepeater
```
Comment (PhilippSpo) [inline src/coffee/client.coffee]: Hm not sure if this was intended, but this condition requires the `enableRepeater` option to exist and to be falsy. This means that if the `enableRepeater` option is not provided it would go into the else block. This looks like an unwanted side-effect to me.

File: src/coffee/repeater-task-queue.coffee [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+      task.reject err
+      return
+    .finally =>
+      @_activeCount -= 1
+      @_maybeExecute()
+    .done()
+
+
+  _shouldRetry: (error) ->
+    if error and (error.code and util.inspect(error.code).startsWith('5') or error.statusCode and util.inspect(error.statusCode).startsWith('5'))
```
Comment (PhilippSpo) [inline src/coffee/repeater-task-queue.coffee]: This is a bit confusing to read.
If I understand correctly `_shouldRetry` should return true if:
- the `error.code` or `error.statusCode` is a 5xx
- the `error` contains one of the retry keywords
I also don't think we need to use nodes `util` library here.

```coffee
_shouldRetry: (error) ->
  return error?.code.toString().startsWith('5')
    or error?.statusCode.toString().startsWith('5')
    or @repeaterOptions.retryKeywords.some (keyword) -> JSON.stringify(error).toUpperCase().includes(keyword.toUpperCase())
```

File: src/coffee/client.coffee (hunk 10 above)
Comment (LEQADA) [inline src/coffee/client.coffee]: @PhilippSpo, as @butenkor [already mentioned](https://github.com/sphereio/sphere-node-sdk/pull/190#discussion_r92011601), the idea is to turn the repeater on if user of SDK will not send any flag to it. So it's not a side-effect, but a feature.

Comment (PhilippSpo) [inline src/coffee/client.coffee]: Hm. True. Then I think this is not the right place to have this default fallback. It should be done like [here](https://github.com/sphereio/sphere-node-sdk/blob/master/src/coffee/client.coffee#L228). But this is a general problem in the SDK right now as far as I can see.

File: src/spec/repeater-task-queue.spec.coffee [hunk 12]
```diff
@@ -33,7 +34,7 @@ describe 'RepeaterTaskQueue', ->
 
     @client.products.fetch()
     .then (res) =>
-      expect(@client._rest.GET.calls.length).toEqual 3
+      expect(@client._rest.GET.calls.length).toEqual 4
```
Comment (PhilippSpo) [inline src/spec/repeater-task-queue.spec.coffee]: Now the title of the test is wrong 🤓
